app/demo-loop/LDemoRun.cc

```diff
@@ -0,0 +1,127 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2021 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file LDemoRun.cc
+//---------------------------------------------------------------------------//
+#include "LDemoRun.hh"
+
+#include "base/CollectionStateStore.hh"
+#include "comm/Logger.hh"
+#include "LDemoParams.hh"
+#include "LDemoInterface.hh"
+#include "LDemoKernel.hh"
+
+using namespace celeritas;
+
+namespace demo_loop
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+template<class P, MemSpace M>
+struct ParamsGetter;
+
+template<class P>
+struct ParamsGetter<P, MemSpace::host>
+{
+    const P& params_;
+
+    auto operator()() const -> decltype(auto)
+    {
+        return params_.host_pointers();
+    }
+};
+
+template<class P>
+struct ParamsGetter<P, MemSpace::device>
+{
+    const P& params_;
+
+    auto operator()() const -> decltype(auto)
+    {
+        return params_.device_pointers();
+    }
+};
+
+template<MemSpace M, class P>
+decltype(auto) get_pointers(const P& params)
+{
+    return ParamsGetter<P, M>{params}();
+}
+
+//---------------------------------------------------------------------------//
+template<MemSpace M>
+ParamsData<Ownership::const_reference, M>
+build_params_refs(const LDemoParams& p)
+{
+    ParamsData<Ownership::const_reference, M> ref;
+    ref.geometry  = get_pointers<M>(*p.geometry);
+    ref.materials = get_pointers<M>(*p.materials);
+    CELER_NOT_IMPLEMENTED("TODO: add remaining references");
+
+    return ref;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Launch interaction kernels for all applicable models.
+ *
+ * For now, just launch *all* the models.
+ */
+void launch_models()
+{
+    CELER_NOT_IMPLEMENTED("TODO: add remaining processes");
+    // Create ModelInteractPointers
+    // Loop over physics models IDs
+    // Invoke `interact`
+    // TODO: for this to work on host, we'll need to template
+    // ModelInterface on MemSpace and overload the `interact`
+    // method on Model to work with device pointers.
+}
+
+//---------------------------------------------------------------------------//
+} // namespace
+
+//---------------------------------------------------------------------------//
+LDemoResult run_gpu(LDemoArgs args)
+{
+    CELER_EXPECT(args);
+
+    // Load all the problem data
+    LDemoParams params = load_params(args);
+
+    // Create param interfaces (TODO unify with sim/TrackInterface )
+    ParamsDeviceRef params_ref = build_params_refs<MemSpace::device>(params);
+
+    // Create states (TODO state store?)
+    StateData<Ownership::value, MemSpace::device> state_storage;
+    resize(&state_storage,
```

Comment:
This feels odd ... shouldn't the 2 operation be combined (i.e. create the collection for the right size in the first step rather than resize)?

Reply:
This is one of those oddities of the Collection/Group model: you can only resize an `Ownership::value`, so you'd have to add a special constructor to the `StateData` that uses `enable_if` to resize its members only when `W == Ownership::value`. 

However, I could add a templated `make_value` that constructs and resizes in a single operation, like the `make_ref` that I added to `CollectionBuilder.hh` in #193 . I wrote `CollectionStateStore` so  that  the above operation gets hidden. I just can't use it here because  RNG and Geo  aren't using Collection objects.